hugo/content/docs/recipes/performance/caches.md

```diff
@@ -0,0 +1,215 @@
+
+```typescript
+import type { ValidationAcceptor, ValidationChecks } from 'langium';
+import type { HelloWorldAstType, Person } from './generated/ast.js';
+import type { HelloWorldServices } from './hello-world-module.js';
+import { InferPublisherService } from './infer-publisher-service.js';
+
+/**
+ * Register custom validation checks.
+ */
+export function registerValidationChecks(services: HelloWorldServices) {
+    const registry = services.validation.ValidationRegistry;
+    const validator = services.validation.HelloWorldValidator;
+    const checks: ValidationChecks<HelloWorldAstType> = {
+        Person: validator.checkPersonIsFromKnownPublisher
+    };
+    registry.register(checks, validator);
+}
+
+/**
+ * Implementation of custom validations.
+ */
+export class HelloWorldValidator {
+    private readonly inferPublisherService: InferPublisherService;
+
+    constructor(services: HelloWorldServices) {
+        this.inferPublisherService = services.utilities.inferPublisherService;
+    }
+
+    checkPersonIsFromKnownPublisher(person: Person, accept: ValidationAcceptor): void {
+        if(this.inferPublisherService.inferPublisher(person) === undefined) {
+            accept('warning', `"${person.name}" is not from a known publisher.`, {
+                node: person
+            });
+        }
+    }
+
+}
+```
+
+### Register the service
+
+Finally, we need to register the service in the module. Go to the `hello-world-module.ts` file and add the following code:
+
+```typescript
+export type HelloWorldAddedServices = {
+    utilities: {
+        inferPublisherService: InferPublisherService
+    },
+    validation: {
+        HelloWorldValidator: HelloWorldValidator
+    }
+}
+//...
+export const HelloWorldModule: Module<HelloWorldServices, PartialLangiumServices & HelloWorldAddedServices> = {
+    utilities: {
+        inferPublisherService: (services) => new CachedInferPublisherService(services)
+    },
+    validation: {
+        //add `services` parameter here
+        HelloWorldValidator: (services) => new HelloWorldValidator(services)
+    }
+};
+```
+
+### Test the result
+
+Start the extension and create a `.hello` file with several persons, like this one:
+
+```plaintext
+person Wonderwoman
+person Spiderman
+person Homer //warning: unknown publisher!!
+person Obelix
+```
+
+## Last words
+
+Caching can improve the performance of your language server. It is especially useful for computations that are expensive to calculate. The `DocumentCache` and `WorkspaceCache` are the most common caches to use. The `ContextCache` is useful if you need to store data for a specific context object. If you only need a key-value store, you can use the `SimpleCache`.
+All of these caches are disposable compared to a simple `Map<K, V>`. If you dispose them by calling `dispose()` the entries will be removed and the memory will be freed.
```

Comment:
Suggestion: Normal maps have a `clear` method (and our caches do as well). The special thing about `dispose` in our caches is that they stop reacting to changes in the workspace once they are disposed. If they are not disposed, the reference to the cache event will stay in the `DocumentBuilder`,

Reply:
Added.